transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,697 @@
+                Poll::Ready(Some(addr)) => Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                    local_addr: ip_to_multiaddr(me.listen_addr.ip(), me.listen_addr.port()),
+                    remote_addr: addr.clone(),
+                    upgrade: Box::pin(WebRTCUpgrade::new(
+                        me.udp_mux.clone(),
+                        me.config.clone(),
+                        addr,
+                    )) as BoxFuture<'static, _>,
+                }))),
+                Poll::Ready(None) => Poll::Ready(None),
+                Poll::Pending => Poll::Pending,
+            };
+        }
+    }
+}
+
+impl WebRTCTransport {
+    async fn do_dial(self, addr: Multiaddr) -> Result<Connection<'static>, Error> {
+        let socket_addr =
+            multiaddr_to_socketaddr(&addr).ok_or_else(|| Error::InvalidMultiaddr(addr.clone()))?;
+        if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
+            return Err(Error::InvalidMultiaddr(addr.clone()));
+        }
+
+        let config = self.config.clone();
+
+        let remote = addr.clone(); // used for logging
+        trace!("dialing address: {:?}", remote);
+
+        let fingerprint = self.cert_fingerprint();
+        let se = build_setting_engine(self.udp_mux.clone(), &socket_addr, &fingerprint);
+        let api = APIBuilder::new().with_setting_engine(se).build();
+
+        let peer_connection = api
+            .new_peer_connection(config)
+            .map_err(Error::WebRTC)
+            .await?;
+
+        // Create a datachannel with label 'data'
+        let data_channel = peer_connection
+            .create_data_channel(
+                "data",
+                Some(RTCDataChannelInit {
+                    negotiated: None,
+                    id: Some(1),
+                    ordered: None,
+                    max_retransmits: None,
+                    max_packet_life_time: None,
+                    protocol: None,
+                }),
+            )
+            .await?;
+
+        let (data_channel_rx, data_channel_tx) = oneshot::channel::<Arc<DetachedDataChannel>>();
+
+        // Wait until the data channel is opened and detach it.
+        let d = Arc::clone(&data_channel);
+        data_channel
+            .on_open(Box::new(move || {
+                debug!("Data channel '{}'-'{}' open.", d.label(), d.id());
+
+                let d2 = Arc::clone(&d);
+                Box::pin(async move {
+                    match d2.detach().await {
+                        Ok(detached) => {
+                            if let Err(_) = data_channel_rx.send(detached) {
+                                error!("data_channel_tx dropped");
+                            }
+                        },
+                        Err(e) => {
+                            error!("Can't detach data channel: {}", e);
+                        },
+                    };
+                })
+            }))
+            .await;
```

Comment:
A nice to trick to avoid the variable shadowing is to make an explicit scope that evaluates to the closure:

```suggestion
        // Wait until the data channel is opened and detach it.
        data_channel
            .on_open({
            	      let data_channel = data_channel.clone();
	              Box::new(move || {
	                  debug!("Data channel '{}'-'{}' open.", data_channel.label(), data_channel.id());
	  
	                  let data_channel = data_channel.clone();
	                  Box::pin(async move {
	                      match data_channel.detach().await {
	                          Ok(detached) => {
	                              if let Err(_) = data_channel_rx.send(detached) {
	                                  error!("data_channel_tx dropped");
	                              }
	                          },
	                          Err(e) => {
	                              error!("Can't detach data channel: {}", e);
	                          },
	                      };
	                  })
	              })
            })
            .await;
```